module/state_synchronization/getter.go

```diff
@@ -0,0 +1,313 @@
+
+		// check blob size
+		blobSize := len(blob.RawData())
+
+		if blobSize > s.maxBlobSize {
+			return nil, &BlobSizeLimitExceededError{blob.Cid()}
+		}
+
+		cache[blob.Cid()] = blob
+
+		if blob.Cid() == target {
+			return blob, nil
+		}
+	}
+
+	// the blob channel may be closed as a result of the context being canceled,
+	// in which case we should return the context error.
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	targetLogger.Debug().Msg("blob not found")
+
+	return nil, &BlobNotFoundError{target}
+}
+
+// getBlobs gets the given CIDs from the blobservice, reassembles the blobs, and deserializes the reassembled data into an object.
+//
+// blobs are fetched from the blobservice and sent over a blob channel as they arrive, to a separate goroutine, which performs the
+// deserialization in a streaming fashion.
+func (s *executionDataGetterImpl) getBlobs(ctx context.Context, cids []cid.Cid, logger zerolog.Logger) (interface{}, uint64, error) {
+	bcr, bs := blobs.OutgoingBlobChannel()
+
+	done := make(chan struct{})
+	var v interface{}
+	var deserializeErr error
+
+	// start deserialization goroutine
+	go func() {
+		defer close(done)
+		defer bcr.Close()
+
+		if v, deserializeErr = s.serializer.Deserialize(bcr); deserializeErr != nil {
+			logger.Debug().Err(deserializeErr).Msg("failed to deserialize execution data")
+		}
+	}()
+
+	retrieveErr := s.retrieveBlobs(ctx, bs, cids, logger)
+
+	<-done
+
+	if retrieveErr != nil && !errors.Is(retrieveErr, blobs.ErrClosedBlobChannel) {
+		return nil, 0, retrieveErr
+	}
+
+	if deserializeErr != nil {
+		return nil, 0, &MalformedDataError{deserializeErr}
+	}
+
+	// TODO: deserialization succeeds even if the blob channel reader has still has unconsumed data, meaning that a malicious actor
+	// could fill the blob tree with lots of unnecessary data by appending it at the end of the serialized data for each level.
+	// It's possible that we could detect this and fail deserialization using something like the following:
+	// https://github.com/onflow/flow-go/blob/bd5320719266b045ae2cac954f6a56e1e79560eb/engine/access/rest/handlers.go#L189-L193
+	// Eventually, we will need to implement validation logic on Verification nodes to slash this behavior.
+
+	return v, bcr.TotalBytesRead(), nil
+}
+
+func (s *executionDataGetterImpl) getExecutionDataRoot(
+	ctx context.Context,
+	blockID flow.Identifier,
+	rootID flow.Identifier,
+	blobGetter network.BlobGetter,
+) (*ExecutionDataRoot, error) {
+	blob, err := blobGetter.GetBlob(ctx, flow.IdToCid(rootID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get execution data root blob: %w", err)
+	}
+
+	v, err := s.serializer.Deserialize(bytes.NewBuffer(blob.RawData()))
+	if err != nil {
+		return nil, fmt.Errorf("failed to deserialize execution data root blob: %w", err)
+	}
+
+	edRoot, ok := v.(*ExecutionDataRoot)
+	if !ok {
+		return nil, fmt.Errorf("execution data root blob is not an ExecutionDataRoot")
+	}
+
+	if edRoot.BlockID != blockID {
```

Comment:
I'm not sure how useful is this validation, which was the main purpose of storing it along with the chunk CIDs into the `ExecutionDataRoot`.

Because, a malicious node could also create a `ExecutionDataRoot` with good blockID and bad CIDs. For example,

An honest node uses adder to add `(blockID1, [C1, C2, C3])`, which returns `rootCIDGood`
A malicious node uses adder to add `(blockID1, [C1, C3, C4])`, which returns `rootCIDBad`. 
Now if a follower node uses getter to get `(blockID1, rootCIDBad)`, it will receive `(blockID1, [C1, C3, C4])`. It will not consider it as invalid, beacuse it is able to get chunks from the bad root CID, and the blockID also matches. 

You might say the follower node will get the blockID and rootCID from the sealed result, so it should have good (blockID, rootCID) pair to begin with, but in this case, a good rootCID will always return `(goodBlockID, [C1, C2, C3])`, so it's impossible for blockID to be different. 

Therefore, I think it's not even necessary to put BlockID into `ExecutionDataRoot`. `ExecutionDataRoot` should just be `[]CID`. It's equally secure as containing BlockID

Reply:
In theory that should be correct, but I think the reason I added the BlockID is just to make sure that every Execution Data Root ID is definitely unique. To be fair, I can't think of any specific scenario that would cause them to not already be unique, even without it.

Another reason is that, somewhere down the line, we may decide to optimistically download Execution Data for blocks before they are even sealed. Then, the Block ID in the Execution Data just serves as a sanity check.

Reply:
It's probably not strictly necessary, but I don't see a super compelling reason not to keep it, and it might make certain things simpler. For example, we can use it to easily know what block an Execution Data corresponds to.

Reply:
OK, in that case, this `MismatchedBlockIDError` should be a fatal error probably a software bug. 

The current error type sounds like as if it's originated from a malicious node, which lead me thought through the above reasoning. 

I'm OK to keep this sentinel error, but better to explain in comments that the caller should ensure the blockID and rootCID are valid ones from sealed results. And `MismatchedBlockIDError` is a fatal error.